api/src/opentrons/protocol_reader/file_hasher.py

```diff
@@ -0,0 +1,26 @@
+"""File hashing utiliy class."""
+
+from typing import Sequence
+import anyio
+import unicodedata
+from hashlib import md5
+from .file_reader_writer import BufferedFile
+
+
+class FileHasher:
+    """Hashing utility class that hashes a combination of protocol and labware files."""
+
+    @staticmethod
+    async def hash(files: Sequence[BufferedFile]) -> str:
+        """Sort and hash a sequence of protocol and labware files."""
+        return await anyio.to_thread.run_sync(_hash_sync, files)
+
+
+def _hash_sync(files: Sequence[BufferedFile]) -> str:
+    sorted_files = sorted(files, key=lambda x: unicodedata.normalize("NFC", x.name))
+    name_content_hasher = md5()
+    for file in sorted_files:
+        name_hash = md5(file.name.encode("utf-8")).digest()
+        contents_hash = md5(file.contents).digest()
+        name_content_hasher.update(name_hash + contents_hash)
+    return name_content_hasher.hexdigest()
```

Comment:
This looks good. Here are a couple of suggested changes:

* Normalize the file name both when we're sorting, *and* when we're hashing. If we expect normalization to be necessary, then it only makes sense to do it for both, not just one or the other, right?
* Leave a comment explaining why we're doing this normalization.

```suggestion
def _hash_sync(files: Sequence[BufferedFile]) -> str:
    def normalized_name(file: BufferedFile) -> str:
        # Normalize the file's name in case there would be differences in its
        # form between the robot's filesystem and the HTTP upload.
        # This is precautionary; we're not actually aware of any differences.
        return unicodedata.normalize("NFC", file.name)

    sorted_files = sorted(files, key=normalized_name)
    name_content_hasher = md5()
    for file in sorted_files:
        name_hash = md5(normalized_name(file).encode("utf-8")).digest()
        contents_hash = md5(file.contents).digest()
        name_content_hasher.update(name_hash + contents_hash)
    return name_content_hasher.hexdigest()
```